### PR TITLE
UX: Move category editing/creation to its own page

### PR DESCRIPTION
**Pull request by pmusaraj**
Screenshots: 

image

image

Plugin test failure is expected, fix in https://github.com/discourse/discourse-restricted-replies/pull/3

### CODE REVIEW

File: app/assets/javascripts/discourse/app/controllers/edit-category.js [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
-  selectedTab: null,
+export default Controller.extend({
+  selectedTab: "general",
   saving: false,
   deleting: false,
   panels: null,
   hiddenTooltip: true,
+  createdCategory: false,
+  expandedMenu: false,
+  mobileView: alias("site.mobileView"),
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/controllers/edit-category.js:17]: I don't think it needs to be an alias, you are not going to write mobileView

File: app/assets/javascripts/discourse/app/routes/discovery-edit-category.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import I18n from "I18n";
+import DiscourseRoute from "discourse/routes/discourse";
+import Category from "discourse/models/category";
+
+export default DiscourseRoute.extend({
+  model(params) {
+    return Category.reloadBySlugPath(params.slug).then((result) => {
+      const record = this.store.createRecord("category", result.category);
+      record.setupGroupsAndPermissions();
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/routes/discovery-edit-category.js:9]: is any of this async ?

File: app/assets/javascripts/discourse/app/routes/discovery-edit-child-category.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import I18n from "I18n";
+import DiscourseRoute from "discourse/routes/discourse";
+import Category from "discourse/models/category";
+
+export default DiscourseRoute.extend({
+  model(params) {
+    return Category.reloadBySlug(params.slug, params.parentSlug).then(
+      (result) => {
+        const record = this.store.createRecord("category", result.category);
+        record.setupGroupsAndPermissions();
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/routes/discovery-edit-child-category.js:10]: same question, is any of this async ?

File: app/assets/javascripts/discourse/app/templates/edit-category.hbs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  {{#each panels as |tab|}}
+    {{component tab selectedTab=selectedTab category=model registerValidator=(action "registerValidator")}}
+  {{/each}}
+
+  <div class="edit-category-footer">
+    {{d-button id="save-category" class="btn-primary" disabled=disabled action=(action "saveCategory") label=saveLabel}}
+
+    {{#if model.can_delete}}
+      {{d-button
+        class="btn-danger pull-right"
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/edit-category.hbs:38]: does pull-right achieves anything here ?

File: app/assets/stylesheets/common/base/edit-category.scss [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+  .edit-category-tab-settings {
+    $category-settings-width: unquote(
+      "min(500px, 90%)"
+    ); // Consistent width makes this all easier to read
+    > section {
+      margin-bottom: 1.5em;
+    }
+
+    input[type="number"] {
```
Comment (jjaffeux) [inline app/assets/stylesheets/common/base/edit-category.scss:72]: if we can use this to drop any direct html element styling that would be cool :+1:

**Comment by jjaffeux** [comment]
@pmusaraj you have a failing test which seems related

**Comment by pmusaraj** [comment]
Yes, the failing test is expected, PR for it already in: https://github.com/discourse/discourse-restricted-replies/pull/3

File: app/assets/javascripts/discourse/app/routes/discovery-edit-category.js (hunk 2 above)
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/routes/discovery-edit-category.js:9]: I don't understand the question, can you rephrase?

File: app/assets/stylesheets/common/base/edit-category.scss (hunk 5 above)
Comment (pmusaraj) [reply app/assets/stylesheets/common/base/edit-category.scss:72]: Not sure I understand this either, would you prefer to target something other than `input[type]`?

File: app/assets/javascripts/discourse/app/routes/discovery-edit-category.js (hunk 2 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/routes/discovery-edit-category.js:9]: Im asking if any of the functions called inside the block are async given you dont wait for it

File: app/assets/stylesheets/common/base/edit-category.scss (hunk 5 above)
Comment (jjaffeux) [reply app/assets/stylesheets/common/base/edit-category.scss:72]: yes should style on class not html element

File: app/assets/javascripts/discourse/app/routes/discovery-edit-category.js (hunk 2 above)
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/routes/discovery-edit-category.js:9]: Ah, thanks. No, none of the functions inside the block are async.

File: app/assets/stylesheets/common/base/edit-category.scss (hunk 5 above)
Comment (pmusaraj) [reply app/assets/stylesheets/common/base/edit-category.scss:72]: Hmm, so, I started doing this, but then I stopped. It feels wrong to do it for just this one small part of the app. 

From a quick search, we use `input[type]` much more than we use classes to target inputs. If we device to go with classes instead, it should be a global refactor, that way it's consistent. 

(This did lead to some cleanup in the adjacent CSS though, that is nice.)

**Comment by pmusaraj** [comment]
Any outstanding issues here @jjaffeux ?

**Comment by jjaffeux** [comment]
> Any outstanding issues here @jjaffeux ?

nope, I think this is stellar work and we should merge it asap and iterate from here!

**Review by jjaffeux — APPROVED** [review]
